docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,76 @@
+# Architecture
+
+This library is under development, and it's primarily designed to fulfil needs of the
+open telemetry exporter and therefore some functionality that would be useful in other
+scenarios is not implemented.
+
+## Idea
+
+Design is heavily influenced by peculiarities of the upstream and downstream systems
+about which I know very little.
+
+### Downstream
+
+The downstream system is DataSet REST API and its API call [addEvents](https://app.scalyr.com/help/api#addEvents).
+It has the following properties:
+
+* `D1` - only one request for each session could be made
+* `D2` - when new version is deployed there is few minutes outage
+* `D3` - maximum payload size for the API is 6MB before compression
+
+### Upstream
+
+The upstream system is [open telemetry collector](https://opentelemetry.io/docs/collector/),
+where the datasetexporter is one of many exporters that could be used
+(on the right side of the image). It has the following properties:
+
+* `U1` - data are passed in batches
+* `U2` - each batch could be either accepted or rejected, if it's accepted it cannot be returned, if it's rejected, it may be retried
+* `U3` - function for accepting batches could be called in parallel (with different batches)
+
+### Consequences
+
+As a consequence of the upstream and downstream system following has to happen (codes
+between brackets points to the previous points):
+
+* `C1` - data has to be grouped to be divided into multiple sessions to increase throughput (`D1`)
```

Comment:
Few points about the bullet points:
1. for bullet points is good practice to keep them consistent - https://grammarist.com/punctuation/bullet-points-usage-rules-examples/ - sections - `Keep a Consistent Structure` and `Rules of Bullet Point Punctuation` - and rest is most likely useful as well
2. I was trying to use structure "it does foo, because of bar"

Reply:
I see the consistency concerns, lets just change the order 
```suggestion
* `C1` - incoming batch of EventsBundles (logs & traces) has to be divided into multiple sessions, in order to increase throughput (`D1`), 
```
my point is to change group -> divide. Remember these bullet points define system constrains, does not describes the implementation.